internal/cluster/certificates.go

```diff
@@ -0,0 +1,371 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/cert"
+	bootstrapv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/cluster-api/util/certs"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	rootOwnerValue = "root:root"
+
+	// TLSKeyDataName is the name of the secret key within the certificate secret of the key
+	TLSKeyDataName = "tls.key"
+
+	// TLSCrtDataName is the name of the secret key within the certificate secret of the certificate
+	TLSCrtDataName = "tls.crt"
+
+	// ClusterCAName is the secret name suffix for apiserver CA
+	ClusterCAName = "ca"
+
+	// EtcdCAName is the secret name suffix for the Etcd CA
+	EtcdCAName = "etcd"
+
+	// ServiceAccountName is the secret name suffix for the Service Account keys
+	ServiceAccountName = "sa"
+
+	// FrontProxyCAName is the secret name suffix for Front Proxy CA
+	FrontProxyCAName = "proxy"
+)
+
+// Certificates are the certificates necessary to bootstrap a cluster.
+type Certificates []*Certificate
+
+// NewCertificates return an initialized but empty set of CA certificates needed to bootstrap a cluster.
+func NewCertificates() Certificates {
+	return Certificates{
+		&Certificate{Name: ClusterCAName},
+		&Certificate{Name: EtcdCAName},
+		&Certificate{Name: ServiceAccountName},
+		&Certificate{Name: FrontProxyCAName},
+	}
+}
+
+// GetCertificateByName returns a certificate by the given name.
+// This could be removed if we use a map instead of a slice to hold certificates, however other code becomes more complex.
+func (c Certificates) GetCertificateByName(name string) *Certificate {
```

Comment:
General question about naming (and ok not to make any changes): the type is Certificates, so we could have GetByName, GenerateCertificates->Generate, SaveGeneratedCertificates-> SaveGenerated, GetOrCreateCertificates-> GetOrCreate?

Reply:
I like the shortened versions better :) will update